src/background/providers/cloudflare.test.ts

```diff
@@ -0,0 +1,334 @@
+    const tokens = [new Token(), new Token()];
+    // @ts-ignore
+    provider.setStoredTokens(tokens);
+    const storedTokens = JSON.parse(mockStorage.store['tokens']);
+    expect(storedTokens).toEqual(tokens.map((token) => token.toString()));
+});
+
+test('getBadgeText', () => {
+    // @ts-ignore
+    const storage: Storage = new StorageMock();
+    const updateIcon = jest.fn();
+    const navigateUrl = jest.fn();
+
+    const provider = new CloudflareProvider(storage, { updateIcon, navigateUrl });
+    const tokens = [new Token(), new Token()];
+    // @ts-ignore
+    provider.setStoredTokens(tokens);
+    // @ts-ignore
+    const text = provider.getBadgeText();
+    expect(text).toBe('2');
+});
+
+describe('issuance', () => {
+    describe('handleBeforeRequest', () => {
+        test('valid request', () => {
+            // @ts-ignore
+            const storage: Storage = new StorageMock();
+            const updateIcon = jest.fn();
+            const navigateUrl = jest.fn();
+
+            const provider = new CloudflareProvider(storage, { updateIcon, navigateUrl });
+            const tokens = [new Token(), new Token(), new Token()];
+            const issue = jest.fn(async () => {
+                return tokens;
+            });
+            // @ts-ignore
+            provider.issue = issue;
+            const url = 'https://captcha.website/?__cf_chl_captcha_tk__=query-param';
+            const details = {
+                method: 'POST',
+                url,
+                requestId: 'xxx',
+                frameId: 1,
+                parentFrameId: 1,
+                tabId: 1,
+                type: 'xmlhttprequest' as chrome.webRequest.ResourceType,
+                timeStamp: 1,
+                requestBody: {
+                    formData: {
+                        ['h-captcha-response']: ['body-param'],
+                    },
+                },
+            };
+            const result = provider.handleBeforeRequest(details);
+            expect(result).toEqual({ cancel: true });
+
+            // Use setTimeout to wait for the issuance to be completed.
+            return new Promise((resolve) =>
+                setTimeout(() => {
+                    expect(issue.mock.calls.length).toBe(1);
+                    expect(issue).toHaveBeenCalledWith(url, {
+                        ['h-captcha-response']: 'body-param',
+                    });
+
+                    expect(navigateUrl.mock.calls.length).toBe(1);
+                    expect(navigateUrl).toHaveBeenCalledWith('https://captcha.website/');
+
+                    // Expect the tokens are added.
+                    // @ts-ignore
+                    const storedTokens = provider.getStoredTokens();
+                    expect(storedTokens.map((token) => token.toString())).toEqual(
+                        tokens.map((token) => token.toString()),
+                    );
+                    resolve(null);
+                }),
+            );
```

Comment:
Here you can use the `await` in front of 
` provider.handleBeforeRequest(details)`
so you can remove the `setTimeout` construction.

```suggestion
            const result = await provider.handleBeforeRequest(details);
            expect(result).toEqual({ cancel: true });

            expect(issue.mock.calls.length).toBe(1);
            expect(issue).toHaveBeenCalledWith(url, {
                ['h-captcha-response']: 'body-param',
            });

            expect(navigateUrl.mock.calls.length).toBe(1);
            expect(navigateUrl).toHaveBeenCalledWith('https://captcha.website/');

            // Expect the tokens are added.
            // @ts-ignore
            const storedTokens = provider.getStoredTokens();
            expect(storedTokens.map((token) => token.toString())).toEqual(
                tokens.map((token) => token.toString()),
            );
```

Reply:
` provider.handleBeforeRequest(details)` doesn't return a Promise, so you can't use `await`

Reply:
it doesn't but I think the effect is to synchronize execution. what happened when you tried?

Reply:
I don't think it has any effect. `await` is designed to be used with `Promise` and `handleBeforeRequest` doesn't return any promise. `setTimeout` is used to add a task to the even loop queue which is supposed to be run after any previously pending tasks. https://developer.mozilla.org/en-US/docs/Web/JavaScript/EventLoop#zero_delays for more info